cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java

```diff
@@ -1145,18 +1304,46 @@ public boolean apply(Privilege input) {
   private void verifyDatasetIsEmpty(NamespaceId namespaceId, String datasetName) throws Exception {
     DataSetManager<KeyValueTable> outTableManager = getDataset(namespaceId.toId(), datasetName);
     KeyValueTable outputTable = outTableManager.get();
-    Assert.assertFalse(outputTable.scan(null, null).hasNext());
+    CloseableIterator<KeyValue<byte[], byte[]>> scanner = outputTable.scan(null, null);
+    Assert.assertFalse(scanner.hasNext());
+    scanner.close();
   }
 
-  private void verifyProgramFailure(Map<String, String> programArgs, ProgramManager programManager)
-    throws TimeoutException, InterruptedException {
+  private void verifyProgramFailure(Map<String, String> programArgs, ProgramManager programManager) throws
+    ExecutionException, InterruptedException, TimeoutException {
     programManager.start(programArgs);
     programManager.waitForFinish(5, TimeUnit.MINUTES);
 
     // verify program history just have failures
     List<RunRecord> history = programManager.getHistory();
-    for (RunRecord runRecord : history) {
-      Assert.assertEquals(ProgramRunStatus.FAILED, runRecord.getStatus());
+    for (final RunRecord runRecord : history) {
+      // the program status might not have stamped as failed instantly after the program finishes so wait for
+      // sometime for it to be marked as failed.
+      try {
+        Tasks.waitFor(ProgramRunStatus.FAILED, new Callable<ProgramRunStatus>() {
```

Comment:
should be able to wait for finish with the specified state, no? ``waitForFinish(int time, TimeUnit unit, ProgramRunStatus status)``. Otherwise we do these checks all over.